File: 1317.convert-integer-to-the-sum-of-two-no-zero-integers.cpp
/*
 * @lc app=leetcode id=1317 lang=cpp
 *
 * [1317] Convert Integer to the Sum of Two No-Zero Integers
 */

#include <bits/stdc++.h>
#include "lassert/lassert.h"

using namespace std;

// @lc code=start
class Solution {
 public:
  vector<int> getNoZeroIntegers(int n) {
    int ret = 1;
    while (ret < n) {
      if (!is_zero_number(ret) && !is_zero_number(n - ret)) {
        return vector<int>{ret, n - ret};
      }
      ret++;
    }
    return vector<int>{ret, n - ret};
  }

 private:
  bool is_zero_number(int n) {
    while (n) {
      if (n % 10 == 0) {
        return true;
      }
      n /= 10;
    }
    return false;
  }
};
// @lc code=end

int main() {
  Solution s;
  vector<int> exp;
  exp = {2, 9};
  ASSERT_EQ(s.getNoZeroIntegers(11), exp);
  return 0;
}

File: 1317.convert-integer-to-the-sum-of-two-no-zero-integers.rs
/*
 * @lc app=leetcode id=1317 lang=rust
 *
 * [1317] Convert Integer to the Sum of Two No-Zero Integers
 */

struct Solution;

#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

impl ListNode {
    #[inline]
    fn new(val: i32) -> Self {
        ListNode { next: None, val }
    }
}

#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<std::rc::Rc<std::cell::RefCell<TreeNode>>>,
    pub right: Option<std::rc::Rc<std::cell::RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}

#[macro_export]
macro_rules! list_node {
    () => {
        None
    };
    ( $x:expr,$( $y:expr ),* ) => {
        {
            let mut head = Some(Box::new(ListNode::new($x)));
            let mut tail = &mut head;
            $(
                let node = Some(Box::new(ListNode::new($y)));
                tail.as_mut().unwrap().next = node;
                tail = &mut tail.as_mut().unwrap().next;
            )*
            head
        }
    };
}

// @lc code=start
impl Solution {
    pub fn get_no_zero_integers(n: i32) -> Vec<i32> {
        for a in 1..n {
            let b = n - a;
            if a.to_string().find('0') == None && b.to_string().find('0') == None {
                return vec![a, b];
            }
        }
        vec![]
    }
}
// @lc code=end

#[test]
fn test1() {
    assert_eq!(Solution::get_no_zero_integers(2), vec![1, 1]);
}

#[test]
fn test2() {
    assert_eq!(Solution::get_no_zero_integers(11), vec![2, 9]);
}
